Multiple webpage refresh crashes Rosbridge server

**Description**
I have installed debian package of rosbridge suite for ROS2 Foxy. It runs fine but if i keep refreshing my webpage, it eventually crashes rosbridge server. In the troubleshoot document it mentions that tornado version mismatch can cause a few problems. Where to check which tornado version goes along with Rosbridge 1.2? Since I have used apt-get install to install the package, I don't think I can use rosdep, right?

Following is the Log I get:
[INFO] [launch]: All log files can be found below /home/aw/.ros/log/2022-09-06-15-32-09-003727-ubuntu-75240
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [rosbridge_websocket-1]: process started with pid [75242]
[INFO] [rosapi_node-2]: process started with pid [75244]
[rosapi_node-2] 1662492729.234910 [0]    python3: using network interface ens33 (udp/192.168.72.130) selected arbitrarily from: ens33, docker0
[rosbridge_websocket-1] 1662492729.268578 [0]    python3: using network interface ens33 (udp/192.168.72.130) selected arbitrarily from: ens33, docker0
[rosbridge_websocket-1] [INFO] [1662492729.311832492] [rosbridge_websocket]: Rosbridge WebSocket server started on port 9090
[rosbridge_websocket-1] [INFO] [1662492734.211482632] [rosbridge_websocket]: Client connected. 1 clients total.
[rosbridge_websocket-1] [INFO] [1662492734.313060972] [rosbridge_websocket]: [Client 8ca8269e-ef1b-4d75-8cbb-0195a3a5c1ff] Subscribed to /ui/fb/full_markers
[rosbridge_websocket-1] [INFO] [1662492734.314190071] [rosbridge_websocket]: [Client 8ca8269e-ef1b-4d75-8cbb-0195a3a5c1ff] Subscribed to /ui/fb/pallet1_markers
[rosbridge_websocket-1] [INFO] [1662492734.315069162] [rosbridge_websocket]: [Client 8ca8269e-ef1b-4d75-8cbb-0195a3a5c1ff] Subscribed to /ui/fb/pallet2_markers
[rosbridge_websocket-1] [INFO] [1662492734.315881053] [rosbridge_websocket]: [Client 8ca8269e-ef1b-4d75-8cbb-0195a3a5c1ff] Subscribed to /ui/fb/pallet3_markers
[rosbridge_websocket-1] [INFO] [1662492734.316826799] [rosbridge_websocket]: [Client 8ca8269e-ef1b-4d75-8cbb-0195a3a5c1ff] Subscribed to /ui/fb/pallet4_markers
[rosbridge_websocket-1] [INFO] [1662492734.317806328] [rosbridge_websocket]: [Client 8ca8269e-ef1b-4d75-8cbb-0195a3a5c1ff] Subscribed to /ui_sim/fb/failed_part
[rosbridge_websocket-1] [INFO] [1662492734.318707581] [rosbridge_websocket]: [Client 8ca8269e-ef1b-4d75-8cbb-0195a3a5c1ff] Subscribed to /ui_sim/fb/status
[rosbridge_websocket-1] [INFO] [1662492734.323567436] [rosbridge_websocket]: [Client 8ca8269e-ef1b-4d75-8cbb-0195a3a5c1ff] Subscribed to /ui_sim/fb/placeable_list
[rosbridge_websocket-1] [INFO] [1662492734.325108644] [rosbridge_websocket]: [Client 8ca8269e-ef1b-4d75-8cbb-0195a3a5c1ff] Subscribed to /ui/fb/robot_connected
[rosbridge_websocket-1] [INFO] [1662492736.087180626] [rosbridge_websocket]: Client disconnected. 0 clients total.
[rosbridge_websocket-1] [INFO] [1662492736.189102910] [rosbridge_websocket]: Client connected. 1 clients total.
[rosbridge_websocket-1] [INFO] [1662492736.252766050] [rosbridge_websocket]: [Client d09319ee-1fe5-4e7f-a24c-455b6b9fcc3b] Subscribed to /ui/fb/full_markers
[rosbridge_websocket-1] [INFO] [1662492736.254386085] [rosbridge_websocket]: [Client d09319ee-1fe5-4e7f-a24c-455b6b9fcc3b] Subscribed to /ui/fb/pallet1_markers
[rosbridge_websocket-1] [INFO] [1662492736.255672828] [rosbridge_websocket]: [Client d09319ee-1fe5-4e7f-a24c-455b6b9fcc3b] Subscribed to /ui/fb/pallet2_markers
[rosbridge_websocket-1] [INFO] [1662492736.257348837] [rosbridge_websocket]: [Client d09319ee-1fe5-4e7f-a24c-455b6b9fcc3b] Subscribed to /ui/fb/pallet3_markers
[rosbridge_websocket-1] [INFO] [1662492736.259179276] [rosbridge_websocket]: [Client d09319ee-1fe5-4e7f-a24c-455b6b9fcc3b] Subscribed to /ui/fb/pallet4_markers
[rosbridge_websocket-1] [INFO] [1662492736.260326568] [rosbridge_websocket]: [Client d09319ee-1fe5-4e7f-a24c-455b6b9fcc3b] Subscribed to /ui/fb/robot_connected
[rosbridge_websocket-1] [INFO] [1662492736.946170828] [rosbridge_websocket]: Client disconnected. 0 clients total.
[rosbridge_websocket-1] [INFO] [1662492737.039601475] [rosbridge_websocket]: Client connected. 1 clients total.
[rosbridge_websocket-1] [INFO] [1662492737.079858959] [rosbridge_websocket]: [Client 81247b43-8c75-4b92-8e74-522d7dcbf712] Subscribed to /ui/fb/full_markers
[rosbridge_websocket-1] [INFO] [1662492737.080645017] [rosbridge_websocket]: [Client 81247b43-8c75-4b92-8e74-522d7dcbf712] Subscribed to /ui/fb/pallet1_markers
[rosbridge_websocket-1] [INFO] [1662492737.081891936] [rosbridge_websocket]: [Client 81247b43-8c75-4b92-8e74-522d7dcbf712] Subscribed to /ui/fb/pallet2_markers
[rosbridge_websocket-1] [INFO] [1662492737.086566073] [rosbridge_websocket]: [Client 81247b43-8c75-4b92-8e74-522d7dcbf712] Subscribed to /ui/fb/pallet3_markers
[rosbridge_websocket-1] [INFO] [1662492737.088225882] [rosbridge_websocket]: [Client 81247b43-8c75-4b92-8e74-522d7dcbf712] Subscribed to /ui/fb/pallet4_markers
[rosbridge_websocket-1] [INFO] [1662492737.089469424] [rosbridge_websocket]: [Client 81247b43-8c75-4b92-8e74-522d7dcbf712] Subscribed to /ui/fb/robot_connected
[rosbridge_websocket-1] [INFO] [1662492737.549564256] [rosbridge_websocket]: Client disconnected. 0 clients total.
[rosbridge_websocket-1] [INFO] [1662492737.651192320] [rosbridge_websocket]: Client connected. 1 clients total.
[rosbridge_websocket-1] [INFO] [1662492737.701295198] [rosbridge_websocket]: [Client 2bd74b31-5861-4085-af86-0ef84579a2d3] Subscribed to /ui/fb/full_markers
[rosbridge_websocket-1] [INFO] [1662492737.702374283] [rosbridge_websocket]: [Client 2bd74b31-5861-4085-af86-0ef84579a2d3] Subscribed to /ui/fb/pallet1_markers
[rosbridge_websocket-1] [INFO] [1662492737.703679360] [rosbridge_websocket]: [Client 2bd74b31-5861-4085-af86-0ef84579a2d3] Subscribed to /ui/fb/pallet2_markers
[rosbridge_websocket-1] [INFO] [1662492737.705692660] [rosbridge_websocket]: [Client 2bd74b31-5861-4085-af86-0ef84579a2d3] Subscribed to /ui/fb/pallet3_markers
[rosbridge_websocket-1] [INFO] [1662492737.707388676] [rosbridge_websocket]: [Client 2bd74b31-5861-4085-af86-0ef84579a2d3] Subscribed to /ui/fb/pallet4_markers
[rosbridge_websocket-1] [INFO] [1662492737.708900319] [rosbridge_websocket]: [Client 2bd74b31-5861-4085-af86-0ef84579a2d3] Subscribed to /ui/fb/robot_connected
[rosbridge_websocket-1] [INFO] [1662492738.064621269] [rosbridge_websocket]: Client disconnected. 0 clients total.
[ERROR] [rosbridge_websocket-1]: process has died [pid 75242, exit code -11, cmd '/opt/ros/foxy/lib/rosbridge_server/rosbridge_websocket --ros-args --params-file /tmp/launch_params_vqse4uzf --params-file /tmp/launch_params_9leetaiu --params-file /tmp/launch_params_jz9gye9r --params-file /tmp/launch_params_4i6_684g --params-file /tmp/launch_params_7ztmwddg --params-file /tmp/launch_params_3us2okw4 --params-file /tmp/launch_params_45p84il2'].



- Library Version: 1.2.0
- ROS Version: Foxy
- Platform / OS: Ubuntu 20.04

--- a/rosbridge_library/src/rosbridge_library/internal/subscribers.py
+++ b/rosbridge_library/src/rosbridge_library/internal/subscribers.py
@@ -31,7 +31,7 @@
 # ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 # POSSIBILITY OF SUCH DAMAGE.
 
-from threading import Lock
+from threading import Lock, RLock
 
 from rclpy.qos import DurabilityPolicy, QoSProfile, ReliabilityPolicy
 from rosbridge_library.internal import ros_loader
@@ -123,7 +123,7 @@ class MultiSubscriber:
         # Create the subscriber and associated member variables
         # Subscriptions is initialized with the current client to start with.
         self.subscriptions = {client_id: callback}
-        self.lock = Lock()
+        self.rlock = RLock()
         self.msg_class = msg_class
         self.node_handle = node_handle
         self.topic = topic
@@ -138,8 +138,11 @@ class MultiSubscriber:
 
     def unregister(self):
         self.node_handle.destroy_subscription(self.subscriber)
-        with self.lock:
+        with self.rlock:
             self.subscriptions.clear()
+            if self.new_subscriber:
+                self.node_handle.destroy_subscription(self.new_subscriber)
+                self.new_subscriber = None
 
     def verify_type(self, msg_type):
         """Verify that the subscriber subscribes to messages of this type.
@@ -165,7 +168,7 @@ class MultiSubscriber:
         messages
 
         """
-        with self.lock:
+        with self.rlock:
             # If the topic is latched, adding a new subscriber will immediately invoke
             # the given callback.
             # In any case, the first message is handled using new_sub_callback,
@@ -183,7 +186,7 @@ class MultiSubscriber:
         client_id -- the ID of the client to unsubscribe
 
         """
-        with self.lock:
+        with self.rlock:
             if client_id in self.new_subscriptions:
                 del self.new_subscriptions[client_id]
             else:
@@ -191,7 +194,7 @@ class MultiSubscriber:
 
     def has_subscribers(self):
         """Return true if there are subscribers"""
-        with self.lock:
+        with self.rlock:
             return len(self.subscriptions) + len(self.new_subscriptions) != 0
 
     def callback(self, msg, callbacks=None):
@@ -208,7 +211,7 @@ class MultiSubscriber:
 
         # Get the callbacks to call
         if not callbacks:
-            with self.lock:
+            with self.rlock:
                 callbacks = self.subscriptions.values()
 
         # Pass the JSON to each of the callbacks
@@ -230,7 +233,7 @@ class MultiSubscriber:
         the subscriptions dictionary is updated with the newly incorporated
         subscriptors.
         """
-        with self.lock:
+        with self.rlock:
             self.callback(msg, self.new_subscriptions.values())
             self.subscriptions.update(self.new_subscriptions)
             self.new_subscriptions = {}

--- a/rosbridge_library/src/rosbridge_library/internal/subscribers.py
+++ b/rosbridge_library/src/rosbridge_library/internal/subscribers.py
@@ -209,18 +209,18 @@ class MultiSubscriber:
         """
         outgoing = OutgoingMessage(msg)
 
-        # Get the callbacks to call
-        if not callbacks:
-            with self.rlock:
-                callbacks = self.subscriptions.values()
+        with self.rlock:
+            callbacks = callbacks if callbacks else self.subscriptions.values()
 
-        # Pass the JSON to each of the callbacks
-        for callback in callbacks:
-            try:
-                callback(outgoing)
-            except Exception as exc:
-                # Do nothing if one particular callback fails except log it
-                self.node_handle.get_logger().error(f"Exception calling subscribe callback: {exc}")
+            # Pass the JSON to each of the callbacks
+            for callback in callbacks:
+                try:
+                    callback(outgoing)
+                except Exception as exc:
+                    # Do nothing if one particular callback fails except log it
+                    self.node_handle.get_logger().error(
+                        f"Exception calling subscribe callback: {exc}"
+                    )
 
     def _new_sub_callback(self, msg):
         """

--- a/rosbridge_library/src/rosbridge_library/internal/subscribers.py
+++ b/rosbridge_library/src/rosbridge_library/internal/subscribers.py
@@ -192,7 +192,7 @@ class MultiSubscriber:
     def has_subscribers(self):
         """Return true if there are subscribers"""
         with self.lock:
-            return len(self.subscriptions) != 0
+            return len(self.subscriptions) + len(self.new_subscriptions) != 0
 
     def callback(self, msg, callbacks=None):
         """Callback for incoming messages on the rclpy subscription.

--- a/rosbridge_library/src/rosbridge_library/internal/subscribers.py
+++ b/rosbridge_library/src/rosbridge_library/internal/subscribers.py
@@ -210,7 +210,7 @@ class MultiSubscriber:
         outgoing = OutgoingMessage(msg)
 
         with self.rlock:
-            callbacks = callbacks if callbacks else self.subscriptions.values()
+            callbacks = callbacks or self.subscriptions.values()
 
             # Pass the JSON to each of the callbacks
             for callback in callbacks:

--- a/rosbridge_server/src/rosbridge_server/websocket_handler.py
+++ b/rosbridge_server/src/rosbridge_server/websocket_handler.py
@@ -208,7 +208,6 @@ class RosbridgeWebSocket(WebSocketHandler):
                 "WebSocketClosedError: Tried to write to a closed websocket",
                 throttle_duration_sec=1.0,
             )
-            raise
         except StreamClosedError:
             cls.node_handle.get_logger().warn(
                 "StreamClosedError: Tried to write to a closed stream",

--- a/rosbridge_server/src/rosbridge_server/websocket_handler.py
+++ b/rosbridge_server/src/rosbridge_server/websocket_handler.py
@@ -208,6 +208,8 @@ class RosbridgeWebSocket(WebSocketHandler):
                 "WebSocketClosedError: Tried to write to a closed websocket",
                 throttle_duration_sec=1.0,
             )
+            # If we end up here, a client has disconnected before its message callback(s) could be removed.
+            # To avoid log spamming, we only log a warning and do not re-raise the exception here.
         except StreamClosedError:
             cls.node_handle.get_logger().warn(
                 "StreamClosedError: Tried to write to a closed stream",

